File: blockchain/add.py
from web3 import Web3
import json

# Connect to Ganache
w3 = Web3(Web3.HTTPProvider('http://127.0.0.1:7545'))

# Verify connection
if not w3.isConnected():
    print("Failed to connect to the Ethereum node")
else:
    print("Connected to the Ethereum node")

# Contract ABI
contract_abi = [
    {
        "constant": False,
        "inputs": [
            {"name": "_name", "type": "string"},
            {"name": "_winner", "type": "string"}
        ],
        "name": "addTournament",
        "outputs": [],
        "payable": False,
        "stateMutability": "nonpayable",
        "type": "function"
    },
    {
        "constant": True,
        "inputs": [],
        "name": "getTournamentCount",
        "outputs": [
            {"name": "", "type": "uint256"}
        ],
        "payable": False,
        "stateMutability": "view",
        "type": "function"
    },
    {
        "constant": True,
        "inputs": [
            {"name": "index", "type": "uint256"}
        ],
        "name": "getTournament",
        "outputs": [
            {"name": "", "type": "string"},
            {"name": "", "type": "string"}
        ],
        "payable": False,
        "stateMutability": "view",
        "type": "function"
    }
]

# Contract address
contract_address = '0x8fe065a44AA43106946577353EfbFdb57da72835'

# Check if contract code exists at the address
code = w3.eth.getCode(contract_address)
if code == b'0x':
    print("No contract deployed at the address")
else:
    print("Contract code exists at the address")

    # Initialize contract
    contract = w3.eth.contract(address=contract_address, abi=contract_abi)

    # Set up the account
    account = w3.eth.accounts[0]

    # Add a tournament
    tx_hash = contract.functions.addTournament("Tournament1", "Winner1").transact({'from': account})
    receipt = w3.eth.waitForTransactionReceipt(tx_hash)

    # Check transaction receipt
    if receipt.status == 1:
        print("Transaction successful!")
    else:
        print("Transaction failed!")

    # Check the tournament count
    try:
        tournament_count = contract.functions.getTournamentCount().call()
        print(f"Tournament Count: {tournament_count}")

        # Fetch and print tournament details
        for index in range(tournament_count):
            tournament = contract.functions.getTournament(index).call()
            print(f"Tournament {index}: Name: {tournament[0]}, Winner: {tournament[1]}")
    except Exception as e:
        print(f"Error fetching tournament details: {e}")

File: docker/solidity/files/blockchain/add.js
const {Web3} = require('web3');
const web3 = new Web3(new Web3.providers.HttpProvider('http://127.0.0.1:7545/'));

const contractAbi = [
    {
        "constant": true,
        "inputs": [
          {
            "name": "",
            "type": "uint256"
          }
        ],
        "name": "tournaments",
        "outputs": [
          {
            "name": "name",
            "type": "string"
          },
          {
            "name": "winner",
            "type": "string"
          }
        ],
        "payable": false,
        "stateMutability": "view",
        "type": "function"
      },
      {
        "constant": false,
        "inputs": [
          {
            "name": "_name",
            "type": "string"
          },
          {
            "name": "_winner",
            "type": "string"
          }
        ],
        "name": "addTournament",
        "outputs": [],
        "payable": false,
        "stateMutability": "nonpayable",
        "type": "function"
      },
      {
        "constant": true,
        "inputs": [],
        "name": "getTournamentCount",
        "outputs": [
          {
            "name": "",
            "type": "uint256"
          }
        ],
        "payable": false,
        "stateMutability": "view",
        "type": "function"
      },
      {
        "constant": true,
        "inputs": [
          {
            "name": "index",
            "type": "uint256"
          }
        ],
        "name": "getTournament",
        "outputs": [
          {
            "name": "",
            "type": "string"
          },
          {
            "name": "",
            "type": "string"
          }
        ],
        "payable": false,
        "stateMutability": "view",
        "type": "function"
      }
    ]

const contractAddress = '0xCABb0706C2E5D5b92119D1865212951dc527CbdD';
const contract = new web3.eth.Contract(contractAbi, contractAddress);

async function main() {
    try {
        const accounts = await web3.eth.getAccounts();
        const account = accounts[0];

        // Specify gas price manually (in wei)
        const gasPrice = '20000000000'; // Example gas price: 20 Gwei

        // Add tournament
        const tx = await contract.methods.addTournament('Tournament1', 'Winner1').send({ from: account, gasPrice });
        console.log('Transaction hash:', tx.transactionHash);
        console.log('Transaction successful!');

        // Get the tournament count
        const count = await contract.methods.getTournamentCount().call();
        console.log('Tournament Count:', count);

        // // Fetch each tournament
        // for (let i = 0; i < count; i++) {
        //     const tournament = await contract.methods.getTournament(i).call();
        //     console.log(`Tournament ${i}: Name: ${tournament[0]}, Winner: ${tournament[1]}`);
        // }
    } catch (error) {
        console.error('Error:', error);
    }
}

main();
